### PR TITLE
mgr/rbd_support: support scheduling long-running background operations

### PR DESCRIPTION
**Pull request by dillaman**
Certain RBD operations might take a long (blocking) amount of time so it would be preferable if we offered a way to schedule these tasks to run in the background. We would expose the ability to manage these background tasks via the MGR's "rbd_support" module via "ceph" CLI commands similar to the following:

    ceph rbd task add flatten <image-spec>
    ceph rbd task add remove <image-spec>
    ceph rbd task add trash remove <image-id-spec>
    ceph rbd task add migration execute <image-spec>
    ceph rbd task add migration commit <image-spec>
    ceph rbd task add migration abort <image-spec>
    ceph rbd task cancel <task-id>
    ceph rbd task list

The `ceph-csi` will be one of the first to utilize these new background tasks to offload its workload and support proper re-starting of in-progress events.

### CODE REVIEW

**Comment by rjfd** [comment]
@dillaman FYI, I opened #29048 yesterday to integrate progress mgr module events in the dashboard.

File: src/test/pybind/test_rbd.py [hunk 1]
```diff
@@ -2165,10 +2208,42 @@ def test_migration(self):
...
+        eq(True, received_callback)
+        remove_image()
+
     def test_migrate_abort(self):
         create_image()
         RBD().migration_prepare(ioctx, image_name, ioctx, image_name, features=63,
                                 order=23, stripe_unit=1<<23, stripe_count=1,
                                 data_pool=None)
         RBD().migration_abort(ioctx, image_name)
-        remove_image()
```
Comment (trociny) [inline src/test/pybind/test_rbd.py]: @dillaman Have you removed `remove_image()` intentionally? It seems it is needed (both here and in `test_migrate_abort_with_progress`).

File: src/test/pybind/test_rbd.py [hunk 2]
```diff
@@ -324,6 +324,17 @@ def test_list():
         image_id = image.id()
     eq([{'id': image_id, 'name': image_name}], list(RBD().list2(ioctx)))
 
+@with_setup(create_image)
+def test_remove_with_progress():
+    received_callback = False
+    def progress_cb(current, total):
+        nonlocal received_callback
+        received_callback = True
+        return 0
```
Comment (trociny) [inline src/test/pybind/test_rbd.py]: It would be nice to have a test that returns a negative error code from the callback.

File: src/test/pybind/test_rbd.py (hunk 1 above)
Comment (dillaman) [reply src/test/pybind/test_rbd.py]: Negative, dropped by accident. Thx.

File: src/pybind/mgr/rbd_support/module.py [hunk 3]
```diff
@@ -41,32 +47,55 @@
...
+TASK_REF_IMAGE_ID = "image_id"
+
+TASK_RETRY_INTERVAL = timedelta(seconds=30)
+
+
+def extract_pool_key(pool_spec):
+    if not pool_spec:
+        return GLOBAL_POOL_KEY
+
+    match = re.match(r'^([^/]+)(?:/([^/]+))?$', pool_spec or '')
```
Comment (trociny) [inline src/pybind/mgr/rbd_support/module.py]: nit: why do you need `or ''` here if you check for `not pool_spec` above?

File: src/pybind/mgr/rbd_support/module.py [hunk 4]
```diff
@@ -483,13 +499,687 @@ def get_perf_counters(self, pool_spec, sort_by):
...
+
+    def append_task(self, task):
+        self.tasks_by_sequence[task.sequence] = task
+        self.tasks_by_id[task.task_id] = task
+
+    def add_task(self, ioctx, action, message, refs):
+        # search for dups and return the original
+        for task_id, task in self.tasks_by_id.items():
+            if task.action == action and task.refs == refs:
+                return json.dumps(to_task.dict())
```
Comment (trociny) [inline src/pybind/mgr/rbd_support/module.py]: did you mean `task.to_dict()` here?

File: src/pybind/mgr/rbd_support/module.py [hunk 5]
```diff
@@ -483,13 +499,687 @@ def get_perf_counters(self, pool_spec, sort_by):
...
+            "perm": "w"
+        },
+        {
+            "cmd": "rbd task migration commit "
+                   "name=image_spec,type=CephString",
+            "desc": "Commit an executed migration asynchronously in the background",
+            "perm": "w"
+        },
+        {
+            "cmd": "rbd task migration abort "
```
Comment (trociny) [inline src/pybind/mgr/rbd_support/module.py]: Just thinking, may be it is better to use `rbd task add <command>`  (i.e. `rbd task add migration abort ...` instead of `rbd task migration abort`) so it would be easier to distinguish between task add commands and other task maintenance commands like `rbd task list` or `rbd task cancel`?

File: src/pybind/mgr/rbd_support/module.py [hunk 6]
```diff
@@ -483,13 +499,687 @@ def get_perf_counters(self, pool_spec, sort_by):
...
+
+                    task.in_progress = False
+                    self.in_progress_task = None
+
+                    self.complete_progress(task)
+                    self.remove_task(ioctx, task)
+
+        except (rados.Error, rbd.Error) as e:
+            self.log.error("execute_task: {}".format(e))
+            self.update_progress(task, 0)
```
Comment (trociny) [inline src/pybind/mgr/rbd_support/module.py]: I am still in process of reading it thus I probably just miss this but how does the user may see if the task completed successfully or with an error?

Comment (dillaman) [reply src/pybind/mgr/rbd_support/module.py]: I agree that this is a definite issue right now. The "ceph progress" command shows all running and completed RBD background tasks, but the current progress API doesn't allow you to update the message. Perhaps I'll tweak that API to support that so that I can record failures for completed commands.

File: src/pybind/mgr/rbd_support/module.py (hunk 5 above)
Comment (dillaman) [reply src/pybind/mgr/rbd_support/module.py]: To be honest, that's how I had originally implemented it but then I yanked it. Since it sounds like it makes sense to you as well, I will re-add `add`.

File: src/pybind/mgr/rbd_support/module.py (hunk 3 above)
Comment (dillaman) [reply src/pybind/mgr/rbd_support/module.py]: I think it was left-over from an earlier implementation.

File: src/pybind/mgr/rbd_support/module.py (hunk 6 above)
Comment (rjfd) [reply src/pybind/mgr/rbd_support/module.py]: @dillaman this issue also popped up when I implemented the integration with ceph-dashboard. 
If you're going to make changes to this part of the progress module, here are the dashboard requirements:
* indication of success/failure (mandatory)
* failure description (in case of failure, optional)

I also plan to do another change to the progress module. Currently the event name is just a description, which is not easily parsable. In the dashboard, our async tasks are always identified by a name of the form `<component_name>/<action>`, e.g., `rbd/create`, `rbd/remove`, etc... I would like to add this kind of "convention" to the progress module so that I can parse the event type, and then I can also process the `event.refs` data structure (since it depends on the event type). Do you have any objections to this?

Comment (dillaman) [reply src/pybind/mgr/rbd_support/module.py]: @rjfd Couldn't you either use the `task_id` or the `refs` field in the `progress` module to stash whatever data you need to link the two?

Comment (dillaman) [reply src/pybind/mgr/rbd_support/module.py]: @rjfd FYI -- it looks like this PR breaks the new dashboard progress service since it runs `dict(ev['refs'])` but I am only storing `["rbd_support"]` in the `refs`. The orchestrator also only stores `["orchestrator"]` in the `refs`

Comment (rjfd) [reply src/pybind/mgr/rbd_support/module.py]: > @rjfd Couldn't you either use the `task_id` or the `refs` field in the `progress` module to stash whatever data you need to link the two?

`task_id` (or `Event.id` inside the progress module) is supposed to be a unique identifier of the task/event, hence is usually initialized with a UUID. What I need is a task `type`, and `refs` depends on the task type, i.e., I can only look inside `refs` and understand what's in there if I know the type of the task.

> it looks like this PR breaks the new dashboard progress service since it runs dict(ev['refs']) but I am only storing ["rbd_support"] in the refs

Could you instead pass more information in there, like we do in the dashboard (https://github.com/ceph/ceph/blob/master/src/pybind/mgr/dashboard/controllers/rbd.py#L255). In dashboard for an `rbd/delete` operation we always pass a dict with `{'pool_name': <pool_name>, 'image_name': <image_name>}`. This way we can precisely pinpoint which RBD image is being deleted and show some feedback to the user through the UI.
For instance, the final outcome that I imagine is a user to start an RBD delete image task in the CLI and see in the dashboard UI the image row of the RBD table to become gray and saying that the image is being deleted.

Comment (rjfd) [reply src/pybind/mgr/rbd_support/module.py]: BTW, in the dashboard `refs` is called `metadata`.

Comment (dillaman) [reply src/pybind/mgr/rbd_support/module.py]: @rjfd It's not just this PR, it's the orchestrator module, the OSD PG recovery progress, etc.

Comment (rjfd) [reply src/pybind/mgr/rbd_support/module.py]: I know :-) It just happened that you were also touching this part.
Maybe I could open a PR with my proposal and we can all discuss it there?

In the meantime, to unblock the dashboard breakage you add this patch:
`'metadata': dict(ev['refs']),` -> `'metadata': dict(ev['refs']) if ev['refs'] and isinstance(ev['refs'][0], tuple) else {},`

Comment (dillaman) [reply src/pybind/mgr/rbd_support/module.py]: @rjfd Sounds good, thx.

Comment (dillaman) [reply src/pybind/mgr/rbd_support/module.py]: @rjfd I appended a couple commits to add `progress` module support for failed events and I now pass a `dict` to the `ev.refs` field.

File: src/librbd/image/RemoveRequest.cc [hunk 7]
```diff
@@ -152,8 +152,10 @@ void RemoveRequest<I>::handle_trim_image(int r) {
   ldout(m_cct, 20) << "r=" << r << dendl;
 
   if (r < 0) {
-    lderr(m_cct) << "warning: failed to remove some object(s): "
+    lderr(m_cct) << "failed to remove some object(s): "
                  << cpp_strerror(r) << dendl;
+    send_close_image(r);
```
Comment (trociny) [inline src/librbd/image/RemoveRequest.cc]: I though it was rather a feature that we didn't abort if failed to remove some objects. I don't remember if I saw a real case when it could be useful though, anyway might this change deserve a separate PR (tracker ticket) to be more visible for users?

Comment (dillaman) [reply src/librbd/image/RemoveRequest.cc]: I can't think of a good reason why we should just leave around data blocks. It was definitely something that was inherited from the original version of remove back when it was in "internal.cc". I can split it out to a separate PR, but I don't think anyone is going to look for it in PR descriptions.

Comment (trociny) [reply src/librbd/image/RemoveRequest.cc]: It was just a suggestion. As you wish.

**Comment by dillaman** [comment]
@rjfd No rush, but when you get a chance, can you review my changes to the `progress`? I just want to ensure we are on the same page for how to record the events to support the dashboard. Right now, here is an example:

```
$ ceph progress json
{
  "events": [],
  "completed": [
    {
      "id": "b586a7a8-3100-4819-9a1c-8dbe08bb7b75",
      "message": "Removing image rbd_tasks/50459f17-f529-4d2e-98de-4798d9c547b7",
      "refs": {
        "origin": "rbd_support",
        "pool_name": "rbd_tasks",
        "pool_namespace": "",
        "image_name": "50459f17-f529-4d2e-98de-4798d9c547b7"
      },
      "started_at": 1563900987.7811384,
      "finished_at": 1563900989.6917968,
      "failed": true,
      "failure_message": "Operation canceled"
    },
    {
      "id": "8f144eaf-deb9-46d9-aad4-1f95746ab8d2",
      "message": "Removing image rbd/image",
      "refs": {
        "origin": "rbd_support",
        "pool_name": "rbd",
        "pool_namespace": "",
        "image_name": "image"
      },
      "started_at": 1563901014.1282082,
      "finished_at": 1563901038.7095985
    }
  ]
}
```

**Comment by dillaman** [comment]
@rjfd ... and if the dashboard wants to translate the `rbd_support` progress events to something it knows about, the `ev.refs["action"]` field now includes the action type [1].

[1] https://github.com/ceph/ceph/pull/29054/commits/01083fe334723579c3089be15d7d809f143f5ced#diff-0eb9731567425c7d4afb3332c908c4fcR67-R72

**Comment by rjfd** [comment]
> @rjfd ... and if the dashboard wants to translate the `rbd_support` progress events to something it knows about, the `ev.refs["action"]` field now includes the action type [1].
> 
> [1] [8774317#diff-0eb9731567425c7d4afb3332c908c4fcR67-R72](https://github.com/ceph/ceph/commit/877431758c3f2fde46f1c0d09b2643d173cc888b#diff-0eb9731567425c7d4afb3332c908c4fcR67-R72)

That's great! This way the problem I had with knowing the "event type" is solved with the pair `refs["origin"], refs["action"]`.

File: qa/workunits/rbd/test_rbd_tasks.sh [hunk 8]
```diff
@@ -0,0 +1,254 @@
...
+cleanup() {
+  ceph osd pool rm ${POOL} ${POOL} --yes-i-really-really-mean-it
+
+  rm -rf ${TEMPDIR}
+}
+
+wait_for() {
+  local TEST_FN=$1
+  shift 1
+  local TEST_FN_ARGS=("$@")
```
Comment (trociny) [inline qa/workunits/rbd/test_rbd_tasks.sh]: @dillaman bashizm
```
zhuzha:~/ceph/ceph.ci/build% ../qa/workunits/rbd/test_rbd_tasks.sh 
+ POOL=rbd_tasks
+ POOL_NS=ns1
../qa/workunits/rbd/test_rbd_tasks.sh: 25: ../qa/workunits/rbd/test_rbd_tasks.sh: Syntax error: "(" unexpected (expecting "}")
```

File: src/pybind/mgr/rbd_support/module.py [hunk 9]
```diff
@@ -483,13 +501,733 @@ def get_perf_counters(self, pool_spec, sort_by):
...
+        except ImportError:
+            # progress module is disabled
+            pass
+
+    def update_progress(self, task, progress):
+        self.log.debug("update_progress: task={}, progress={}".format(str(task), progress))
+        try:
+            self.module.remote("progress", "update", task.task_id,
+                               task.message, progress,
+                               {**{"origin": "rbd_support"}, **task.refs})
```
Comment (trociny) [inline src/pybind/mgr/rbd_support/module.py]: Do we still support python2?

I happened to try this on Ubuntu 16.04, built with python2.7, and observing this error:
```
SyntaxError: ('invalid syntax', ('/home/mgolub/ceph/ceph.ci/src/pybind/mgr/rbd_support/module.py', 923, 34, '                               {**{"origin": "rbd_support"}, **task.refs})\n'))
```
which I believe due to python2.7?

File: qa/workunits/rbd/test_rbd_tasks.sh (hunk 8 above)
Comment (dillaman) [reply qa/workunits/rbd/test_rbd_tasks.sh]: Thx -- explicitly use bash now

File: src/pybind/mgr/rbd_support/module.py (hunk 9 above)
Comment (dillaman) [reply src/pybind/mgr/rbd_support/module.py]: Switch this to a `{}.update()` call

File: qa/workunits/rbd/test_rbd_tasks.sh [hunk 10]
```diff
@@ -0,0 +1,254 @@
+#!/bin/bash -ex
```
Comment (trociny) [inline qa/workunits/rbd/test_rbd_tasks.sh]: @dillaman The preferred way is to use:
```
#!/usr/bin/env bash
set -ex
```
Because on some platforms (FreeBSD) bash is not installed to `/bin`.

File: src/pybind/mgr/rbd_support/module.py [hunk 11]
```diff
@@ -483,13 +501,735 @@ def get_perf_counters(self, pool_spec, sort_by):
...
+        if image_spec[1]:
+            image = "{}/{}".format(image_spec[1], image)
+        if image_spec[0]:
+            image = "{}/{}".format(image_spec[0], image)
+        return image
+
+    def init_task_queue(self):
+        for pool_id, pool_name in get_rbd_pools(self.module).items():
+            try:
+                with self.module.rados.open_ioctx2(pool_id) as ioctx:
```
Comment (trociny) [inline src/pybind/mgr/rbd_support/module.py]: By some reason python2 does not think `pool_id` is int here:
```
2019-07-25T11:23:00.522+0300 7ff87e52d700 -1 mgr load Traceback (most recent call last):
  File "/home/mgolub/ceph/ceph.ci/src/pybind/mgr/rbd_support/module.py", line 1207, in __init__
    self.task = TaskHandler(self)
  File "/home/mgolub/ceph/ceph.ci/src/pybind/mgr/rbd_support/module.py", line 607, in __init__
    self.init_task_queue()
  File "/home/mgolub/ceph/ceph.ci/src/pybind/mgr/rbd_support/module.py", line 668, in init_task_queue
    with self.module.rados.open_ioctx2(pool_id) as ioctx:
  File "rados.pyx", line 564, in rados.requires.wrapper.validate_func (/home/mgolub/ceph/ceph.ci/build/src/pybind/rados2/rados.c:8341)
  File "rados.pyx", line 552, in rados.requires.check_type (/home/mgolub/ceph/ceph.ci/build/src/pybind/rados2/rados.c:7732)
TypeError: pool_id must be int
```
`self.module.rados.open_ioctx2(int(pool_id))` fixes this.

File: src/pybind/mgr/rbd_support/module.py [hunk 12]
```diff
@@ -483,13 +501,735 @@ def get_perf_counters(self, pool_spec, sort_by):
...
+        if pool_spec == GLOBAL_POOL_KEY:
+            pool_spec = (self.default_pool_name, '')
+        return pool_spec
+
+    def extract_image_spec(self, image_spec):
+        match = re.match(r'^(?:([^/]+)/(?:([^/]+)/)?)?([^/@]+)$',
+                         image_spec or '')
+        if not match:
+            raise ValueError("Invalid image spec: {}".format(image_spec))
+        return (match[1] or self.default_pool_name, match[2] or '', match[3])
```
Comment (trociny) [inline src/pybind/mgr/rbd_support/module.py]: python2 does not know how to access match results by `[]`:
```
Error EINVAL: Traceback (most recent call last):
  File "/home/mgolub/ceph/ceph.ci/src/pybind/mgr/mgr_module.py", line 914, in _handle_command
    return self.handle_command(inbuf, cmd)
  File "/home/mgolub/ceph/ceph.ci/src/pybind/mgr/rbd_support/module.py", line 1218, in handle_command
    return self.task.handle_command(inbuf, prefix[9:], cmd)
  File "/home/mgolub/ceph/ceph.ci/src/pybind/mgr/rbd_support/module.py", line 1113, in handle_command
    return self.queue_remove(cmd['image_spec'])
  File "/home/mgolub/ceph/ceph.ci/src/pybind/mgr/rbd_support/module.py", line 957, in queue_remove
    image_spec = self.extract_image_spec(image_spec)
  File "/home/mgolub/ceph/ceph.ci/src/pybind/mgr/rbd_support/module.py", line 627, in extract_image_spec
    return (match[1] or self.default_pool_name, match[2] or '', match[3])
TypeError: '_sre.SRE_Match' object has no attribute '__getitem__'
```
I had to change to `match.group(1)` (here and in all other places).

File: src/pybind/mgr/rbd_support/module.py [hunk 13]
```diff
@@ -483,13 +501,735 @@ def get_perf_counters(self, pool_spec, sort_by):
...
+        self.log.debug("progress_callback: task={}, progress={}".format(
+            str(task), progress))
+
+        # avoid deadlocking when a new command comes in during a progress callback
+        if not self.lock.acquire(False):
+            return 0
+
+        try:
+            if not self.in_progress_task or self.in_progress_task.canceled:
+                return -errno.ECANCELED
```
Comment (trociny) [inline src/pybind/mgr/rbd_support/module.py]: python2 does not knows about `errno.ECANCELED`. Due to the exception was catched, the error did not propagate and `progeress_callback` returned 0, so cancel did not work. To proceed with testing I changed it to -125 locally.

File: src/test/pybind/test_rbd.py [hunk 14]
```diff
@@ -324,6 +325,25 @@ def test_list():
         image_id = image.id()
     eq([{'id': image_id, 'name': image_name}], list(RBD().list2(ioctx)))
 
+@with_setup(create_image)
+def test_remove_with_progress():
+    received_callback = False
+    def progress_cb(current, total):
+        nonlocal received_callback
```
Comment (trociny) [inline src/test/pybind/test_rbd.py]: python2 does not recognize `nonlocal`

File: src/pybind/mgr/rbd_support/module.py [hunk 15]
```diff
@@ -483,13 +501,714 @@ def get_perf_counters(self, pool_spec, sort_by):
...
+                ex, traceback.format_exc()))
+
+    @contextmanager
+    def open_ioctx(self, spec):
+        try:
+            with self.module.rados.open_ioctx(spec[0]) as ioctx:
+                ioctx.set_namespace(spec[1])
+                yield ioctx
+        except rados.ObjectNotFound:
+            self.log.error("Failed to locate pool {}".format(spec[0]))
```
Comment (trociny) [inline src/pybind/mgr/rbd_support/module.py]: Shouldn't we propagate the exception in this case?
When the test failed and pool was removed while the task were still in proceed I was observing this:
```
2019-07-25T12:13:24.143+0300 7f58e04d9700  0 mgr[rbd_support] Failed to locate pool rbd_tasks
2019-07-25T12:13:24.439+0300 7f58e04d9700  0 mgr[rbd_support] Fatal runtime error: generator didn't yield
Traceback (most recent call last):
  File "/home/mgolub/ceph/ceph.ci/src/pybind/mgr/rbd_support/module.py", line 638, in run
    self.execute_task(sequence)
  File "/home/mgolub/ceph/ceph.ci/src/pybind/mgr/rbd_support/module.py", line 779, in execute_task
    task.refs[TASK_REF_POOL_NAMESPACE])) as ioctx:
  File "/usr/lib/python2.7/contextlib.py", line 19, in __enter__
    raise RuntimeError("generator didn't yield")
RuntimeError: generator didn't yield
```

**Comment by trociny** [comment]
And yet another thing noticed. When the test failed for some reason while there were tasks in the mgr in-memory queue, and then the pool was deleted, the task remained int the mgr in-memory queue until it was restarted.

File: src/pybind/mgr/rbd_support/module.py (hunk 13 above)
Comment (dillaman) [reply src/pybind/mgr/rbd_support/module.py]: Since I already had to add `ECANCELED` to `rbd.pyx`, I think I'll just re-export it so that I don't need to hard-code a constant.

File: src/pybind/mgr/rbd_support/module.py (hunk 15 above)
Comment (dillaman) [reply src/pybind/mgr/rbd_support/module.py]: Exception re-thrown and now the `execute_task` catches the exception to delete the task if the pool is missing.

File: src/pybind/mgr/rbd_support/module.py (hunk 12 above)
Comment (dillaman) [reply src/pybind/mgr/rbd_support/module.py]: Done, thx

File: src/pybind/mgr/rbd_support/module.py (hunk 11 above)
Comment (dillaman) [reply src/pybind/mgr/rbd_support/module.py]: Fixed

**Comment by dillaman** [comment]
The new `rbd_tasks` teuthology test is still failing [1]. Will investigate.

[1] http://pulpito.ceph.com/jdillaman-2019-07-25_10:51:34-rbd-wip-jd-testing-distro-basic-smithi/

**Comment by dillaman** [comment]
Teuthology test is now passing:

http://pulpito.ceph.com/jdillaman-2019-07-25_15:24:54-rbd-wip-jd-testing-distro-basic-smithi/
http://pulpito.ceph.com/jdillaman-2019-07-25_15:25:24-rbd-wip-jd-testing-distro-basic-smithi/

File: src/pybind/mgr/rbd_support/module.py [hunk 16]
```diff
@@ -483,13 +502,802 @@ def get_perf_counters(self, pool_spec, sort_by):
...
+            self.remove_task(ioctx, task, remove_in_memory)
+
+        return 0, "", ""
+
+    def task_list(self, task_id):
+        self.log.info("task_list: {}".format(task_id))
+
+        if task_id:
+            if task_id not in self.tasks_by_id:
+                raise KeyError("No such task {}".format(task_id))
```
Comment (trociny) [inline src/pybind/mgr/rbd_support/module.py]: Should we rise an exception here? Because it logs like below:
```
2019-07-26T11:10:57.208+0300 7fb047494700  0 log_channel(audit) log [DBG] : from='client.4694 -' entity='client.admin' cmd=[{"prefix": "rbd task list", "target": ["mgr", ""], "task_id": "e87e6984-8953-400d-8860-31dfd22764c8"}]: dispatch
2019-07-26T11:10:57.208+0300 7fb047c95700  0 mgr[rbd_support] Fatal runtime error: 'No such task e87e6984-8953-400d-8860-31dfd22764c8'
Traceback (most recent call last):
  File "/home/mgolub/ceph/ceph.ci/src/pybind/mgr/rbd_support/module.py", line 1284, in handle_command
    return self.task.handle_command(inbuf, prefix[9:], cmd)
  File "/home/mgolub/ceph/ceph.ci/src/pybind/mgr/rbd_support/module.py", line 1191, in handle_command
    return self.task_list(cmd.get('task_id'))
  File "/home/mgolub/ceph/ceph.ci/src/pybind/mgr/rbd_support/module.py", line 1163, in task_list
    raise KeyError("No such task {}".format(task_id))
KeyError: 'No such task e87e6984-8953-400d-8860-31dfd22764c8'
```
which does not look very nice. I am not sure it even should be logged as an error, I would log it as debug.

The same is probably for `task_cancel`.

**Comment by trociny** [comment]
And just as note, I suppose this this a bug in progress module, but when displaying a task progress (either with `ceph progress` or `ceph status`) the output looks like below:
```
  progress:
    Removing image test_task/XXXMG (since 00h 00m 54s)
      [========================......]
```
And this `since 00h 00m 54s` looks confusing to me. Shouldn't it show the task start time? Or does it mean task duration? Then I would use e.g. `for 00h 00m 54s` , not `since ...`.

**Comment by rjfd** [comment]
> And this `since 00h 00m 54s` looks confusing to me. Shouldn't it show the task start time? Or does it mean task duration? Then I would use e.g. `for 00h 00m 54s` , not `since ...`.

Yes, it's the duration. And I agree "since ..." is misleading.

**Comment by dillaman** [comment]
I added a tracker ticket [1] for the unrelated progress issue.

[1] http://tracker.ceph.com/issues/40976

File: src/pybind/mgr/rbd_support/module.py (hunk 16 above)
Comment (dillaman) [reply src/pybind/mgr/rbd_support/module.py]: Updated

**Comment by trociny** [comment]
Just cosmetic. When remove is canceled, we have in the mgr log:
```
2019-07-26T14:23:17.338+0300 7f59a97d1700 -1 librbd::TrimRequest: trim encountered an error: (125) Operation canceled
2019-07-26T14:23:17.338+0300 7f59a8fd0700 -1 librbd::image::RemoveRequest: 0x56096bf6f1e0 handle_trim_image: failed to remove some object(s): (125) Operation canceled
```
It does not look very bad, as I suppose for a user from the error message it is clear that it is not actually an error but just cancel, but I can imagine it could be improved if you think it is worth doing.

**Comment by dillaman** [comment]
> It does not look very bad, as I suppose for a user from the error message it is clear that it is not actually an error but just cancel, but I can imagine it could be improved if you think it is worth doing.

It could be improved in a later PR I think (throughout since it will affect more than just removal I'd suspect).
